tests/installation/logs_from_installation_system.pm

```diff
@@ -0,0 +1,48 @@
+# SUSE's openQA tests
+#
+# Copyright © 2009-2013 Bernhard M. Wiedemann
+# Copyright © 2012-2017 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Collect logs from the installation system just before we try to
+#   reboot into the installed system
+# Maintainer: Oliver Kurz <okurz@suse.de>
+
+use strict;
+use base "y2logsstep";
+use testapi;
+use lockapi;
+use utils;
+use ipmi_backend_utils;
+
+sub run {
+    return if get_var("REMOTE_CONTROLLER") || is_caasp || is_hyperv_in_gui;
+    select_console 'install-shell';
+
+    # check for right boot-device on s390x (zVM, DASD ONLY)
+    if (check_var('BACKEND', 's390x') && !check_var('S390_DISK', 'ZFCP')) {
+        if (script_run('lsreipl | grep 0.0.0150')) {
+            die "IPL device was not set correctly";
+        }
+    }
+    # while technically SUT has a different network than the BMC
+    # we require ssh installation anyway
+    if (check_var('BACKEND', 'ipmi')) {
+        use_ssh_serial_console;
+        # set serial console for xen
+        set_serial_console_on_xen("/mnt") if (get_var("XEN") || check_var("HOST_HYPERVISOR", "xen"));
+    }
+    else {
+        # avoid known issue in FIPS mode: bsc#985969
+        $self->get_ip_address();
+    }
+    $self->save_upload_y2logs();
+    select_console 'installation';
```

Comment:
It's arguable if this select console should not rather be in the reboot_after_installation.pm - with the modular approach, more things can potentially appear in between

Reply:
Hm, the later use case you describe is actually one of the things I would like to enable but OTOH this test module does first `select_console 'install-shell';` so it should clean up after itself, therefore go back where it came from. So I see the following options:
1) Do the switch to "install-shell" already in "await_install", the switch back in "reboot_after_installation". *Drawback:* The test modules are not following their contract but something which is necessary only for another module or other modules. *Benefit:* Only one switch back and forth
2) Explicit test modules for switching. *Drawback:* Two more test modules for every single test run when in most cases we don't want to do anything more than just the standard single "logs_from_installation_system". *Benefit:* Very explicit, only single switches
3) Every intermediate test module would do the corresponding console switch. *Drawback:* We switch multiple times for every optional module, *Benefit:* It's more robust. The default workflow stays very simple, only optional modules would introduce additional switching

Reply:
one could also argue that a module should ensure it is on the console it expects to be on... select_console FOO does not perform a switch if FOO is already selected

Reply:
correct. But that doesn't change that you wanted to *prevent* more switches than necessary in the first place, isn't it?

Reply:
if the test in the end 'remains where it is' and not 'where it believes it was from when the test started', yes it does. So in short: never switch console as last action. That only adds an assumption of what will come after, but you can't ever be sure this to remain true

Reply:
you are right. I moved the `select_console 'installation';` to `installation/reboot_after_installation.pm`